File: FishEyeDemo/FishEyeDemo/ModelPoint.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace FishEyeDemo
{
    class ModelPoint
    {
        public double x;
        public double y;

        public ModelPoint(double x, double y)
        {
            this.x = x;
            this.y = y;
        }

        public bool Equals(ModelPoint other, double tolerance = 0)
        {
            return (Math.Abs(this.x - other.x) <= tolerance && Math.Abs(this.y - other.y) <= tolerance);
        }

        public double Dot(ModelPoint other)
        {
            return this.x * other.x + this.y * other.y;
        }

        public double LengthCross(ModelPoint other)
        {
            return this.x * other.y - this.y * other.x;
        }

        public double TurnAngle(ModelPoint other1, ModelPoint other2)
        {
            ModelPoint v1 = other1 - this;
            ModelPoint v2 = other2 - other1;
            double length_cross = v1.LengthCross(v2);
            double dot = v1.Dot(v2);
            if (length_cross >= 0)
            {
                return Math.Acos(length_cross / v1.Length() / v2.Length());
            }
            else
            {
                return -Math.Acos(length_cross / v1.Length() / v2.Length());
            }
        }

        public double Length()
        {
            return Math.Sqrt(Math.Pow(this.x, 2) + Math.Pow(this.y, 2));
        }

        public void Normalize()
        {
            double length = this.Length();
            x /= length;
            y /= length;
        }

        public static ModelPoint operator -(ModelPoint p)
        {
            return new ModelPoint(-p.x, -p.y);
        }

        public static ModelPoint operator +(ModelPoint p1, ModelPoint p2)
        {
            return new ModelPoint(p1.x + p2.x, p1.y + p2.y);
        }

        public static ModelPoint operator -(ModelPoint p1, ModelPoint p2)
        {
            return new ModelPoint(p1.x - p2.x, p1.y - p2.y);
        }
    }
}

File: MetaServer/project/static/javascript/modelpoint.js
function ModelPoint(x, y) {
	this.x = x;
	this.y = y;
	
	this.Equals = function(otherModelPoint, tolerance = 0) {
		return (Math.abs(this.x - otherModelPoint.x) <= tolerance && Math.abs(this.y - otherModelPoint.y) <= tolerance);
	};
	
	this.Dot = function(otherModelPoint) {
			return this.x * otherModelPoint.x + this.y * otherModelPoint.y;
	};
	
	this.LengthCross = function(otherModelPoint) {
			return this.x * otherModelPoint.y - this.y * otherModelPoint.x;
	};
	
	this.TurnAngle = function(otherModelPoint1, otherModelPoint2) {
			v1 = otherModelPoint1.Subtract(this);
			v2 = otherModelPoint2.Subtract(otherModelPoint1);
			length_cross = v1.LengthCross(v2);
			dot = v1.Dot(v2);
			if (dot >= 0)
			{
				return Math.asin(length_cross / v1.Length() / v2.Length());
			}
			else
			{
				sgn = Math.sign(length_cross) >= 0 ? 1 : -1;
				return sgn * Math.PI - Math.asin(length_cross / v1.Length() / v2.Length());
			}
	};

	this.InTriangle = function(modelPointA, modelPointB, modelPointC)
	{
		modelPointAB = modelPointB - modelPointA;
		modelPointBC = modelPointC - modelPointB;
		modelPointCA = modelPointA - modelPointC;
		sign1 = ab.LengthCross(this - modelPointA);
		sign2 = bc.LengthCross(this - modelPointB);
		sign3 = ca.LengthCross(this - modelPointC);
		return (sign1 >= 0) && (sign2 >= 0) && (sign3 >= 0) || (sign1 <= 0) && (sign2 <= 0) && (sign3 <= 0); 
	}
	
	this.Length = function()
	{
		return Math.sqrt(Math.pow(this.x, 2) + Math.pow(this.y, 2));
	};
	
	this.Negate = function()
	{
		return new ModelPoint(-this.x, -this.y);
	};

	this.Add = function(otherModelPoint)
	{
		return new ModelPoint(this.x + otherModelPoint.x, this.y + otherModelPoint.y);
	};

	this.Subtract = function(otherModelPoint)
	{
		return new ModelPoint(this.x - otherModelPoint.x, this.y - otherModelPoint.y);
	};

	
	this.toString = function() {
		return `(${this.x};${this.y})`;
	};
}

export default ModelPoint;
